충남대 BE_배민수 4주차 과제 (2단계)

minsu-cnu opened the pull request:
1단계 리뷰에 대한 답을 1단계 PR에 답변 남겨두었습니다!

그리고 1단계 리뷰 중에서 코드 변경에 관한 내용은 반영하여 해당 PR의 커밋에 포함하였습니다.

<br/>

1단계 리뷰 남겨주신 내용에 대해 한 가지 추가적인 질문이 있어 답변 남겨두었는데, 여기에도 다시 작성하겠습니다!

<br/>
<br/>

## 리뷰 내용

가만보니 이제 mock을 사용하는 테스트가 없군요!

테스트에서 어떤 방법을 선호 하는지는 아직도 많은 얘기가 오가는 주제인데요
Classicist vs Mockist 중에서 민수님은 Classicist 쪽인 것 같네요!ㅎㅎ
어떤 점에서 장점을 느꼈을까요?

<br/>

## 답변 및 추가 질문

사실 service test에서, 예를 들어서 WishServiceTest에서는 스프링 컨테이너를 띄워서 WishService를 주입시키고 있는데, 이를 속도가 빠른 단위 테스트로 바꿔보고 싶어서, 해당 테스트에서 의존하는 ProductRepository와 CategoryRepository를 mock 객체로 만들고, WishService는 new 연산자로 직접 객체를 생성해서 사용하도록 리팩토링해보고자 했었습니다.

그런데 WishService를 new 연산자로 직접 생성해서 쓰게 되면, 만약 다른 테스트에서 스프링 빈으로 등록된 WishService를 사용하게 되면, 이 스프링 빈 객체와는 다른 WishService 객체를 생성해서 쓰게 되는 것이기 때문에 예상치 못한 사이드 이펙트가 생길 수도 있을 것 같다는 걱정이 들기도 하고, 시간도 빠듯해서 적용을 하지는 않았었습니다.

혹시 해당 방식으로 단위 테스트로 바꾸어도 문제는 없을지, 적절한 방식인지 궁금합니다!

한편, mock을 사용하지 않는 방식의 장점으로는, 제가 느낀 바로는 우선 mock 객체를 사용하는 코드에 비해 가독성이 조금 더 좋은 것 같고, 특히 mock 방식의 경우 행위를 일일히 지정해줘야하는데, 만약 코드가 변경된다면 이 것이 깨지게 되어서 유지 보수가 번거로울 수 있기 때문에, 구현체를 사용하는 방식이 더 좋은 상황도 있는 것 같습니다.

mock의 단점을 보완하기 위한 방식으로 구현체를 사용하는 방식도 있고, fake 객체를 사용하는 방법도 있는 것 같습니다. fake 객체를 따로 또 구현해줘야한다는 번거로움이 있긴 하지만 장점도 있기 때문에, 상황에 맞게 위 3가지 방식 중 알맞은 것을 골라서 쓸 수 있는 능력이 중요할 것 같습니다!

minsu-cnu committed:
feat: insert 요청 수행 시 201 created 상태 코드를 응답하도록 변경

minsu-cnu committed:
test: isPresent 사용 리팩토링

minsu-cnu committed:
docs: 2단계 기능 명세

minsu-cnu committed:
docs: 2단계 기능 명세 수정

minsu-cnu committed:
test: 불필요한 AfterEach 삭제

minsu-cnu committed:
test: 테스트 클래스명 수정

minsu-cnu committed:
test: 불필요한 test instance 생명 주기 설정 삭제

minsu-cnu committed:
feat: 옵션 엔티티 도메인 클래스, 요청 dto, 응답 dto 구현

minsu-cnu committed:
feat: 옵션 컨트롤러, 서비스, 레포지토리 구현

minsu-cnu committed:
fix: 옵션 컨트롤러 요청 dto에 valid 어노테이션 누락된 것 추가

minsu-cnu committed:
test: 요청 dto 유효성 검증 예외 처리 테스트 추가

minsu-cnu committed:
style: 자바 코드 컨벤션 적용

jjongwa commented on src/main/java/gift/product/model/Option.java:
```diff
@@ -0,0 +1,54 @@
+    @Column(nullable = false)
+    private final int quantity;
+
+    @ManyToOne
+    @JoinColumn(nullable = false)
+    private final Product product;
```
단방향 관계를 선택한 이유는 무엇인가요?

옵션 정보를 상품 없이 조회하는 경우가 있을까요?

minsu-cnu replied:
단방향 관계를 선택한 근거는 다음과 같습니다.

기능 요구 사항에서, 클라이언트가 상품 ID를 지정하여 그 상품에 대한 옵션을 응답받는 API를 구현할 것을 요구하고 있습니다. 이 것을 보고 저는 클라이언트가 상품 ID를 알고 있는 상태에서 그 상품에 대한 옵션을 요청하기 때문에, 서버에서는 상품 객체를 통해 연관된 옵션 리스트를 참조하지 않고도, 충분히 상품 ID를 통해 옵션 레포지토리에서 연관 옵션들을 찾아낼 수 있으므로, 양방향 관계를 설정하지 않았습니다.

단방향 관계를 설정한 또 다른 이유 중 하나는, 관리자 입장에서 통계나 어떠한 비즈니스 결정을 위해, 관리자 페이지 등에서 플랫폼에 존재하는 모든 옵션을 상품에 관계없이 전체 조회를 하고 싶은 경우가 있을 수도 있겠다는 생각이 있었기 때문입니다.

그런데 다시 생각해보면 단방향 관계와 양방향 관계에 각각 장단점이 있는 것 같습니다.

현재 구현 방식인 단방향 관계의 경우, 양방향 관계를 설정하지 않아도 되는 장점이 있지만, 상품 ID로 옵션 테이블을 전부 조회해서 연관된 옵션들을 찾아내는 작업이 필요해서 비용이 큰 것 같습니다.

양방향 관계의 경우, 상품 ID로 상품 객체만 찾아내면, 해당 객체가 참조하고 있는 옵션 객체 리스트를 그대로 가져가면 되기 때문에 객체 중심의 JPA 특성을 잘 살릴 수 있고, DB 연산 측면에서 성능 및 비용 상 이점이 있는 것 같습니다. 다만 양뱡향 관계를 설정할 경우 복잡성이 증가하고, 상품을 로드할 때 연관된 옵션을 모두 로드하는 등 경우에 따라 성능 상 이슈가 있을 듯 합니다.

이 두 가지 방식 중 trade-off의 정도를 비교하여 방식을 정해야하는 것 같은데 어떤 방식이 현재 상황에 최적인지 감이 잘 잡히지 않습니다. 혹시 이에 관한 좋은 의견이 있으시다면 조언 주시면 감사드리겠습니다!

jjongwa replied:
이런 부분은 어떠한 정답이 있는게 아니라서 이렇게 본인의 의견과 이렇게 생각한 이유를 정리하는 연습을 하는 점에 중점을 두면 좋을 것 같아요! 

지금은 저 한명만 코드리뷰를 진행하고 있지만 현업이나 팀플을 하게 되면 여러 명이 서로 반대되는 리뷰를 달 수도 있습니다! 그러한 상황에서 각자의 방식을 잘 이해하고 최적의 선택을 하는게 중요해요~!

jjongwa reviewed (changes requested):
현재 모든 테스트를 한번에 실행 시 실패하는 테스트가 발생하고 있어요!
돌릴 때마다 실패하는 테스트의 수가 변하는데, 어떤 문제 때문일지 알고 계실 것 같네요..!

<img width="495" alt="image" src="https://github.com/user-attachments/assets/c44f9cbd-b1e7-4cbd-8c5e-e0c2da602140">
<img width="560" alt="image" src="https://github.com/user-attachments/assets/c8cf3076-d077-4a9c-ac11-adad985b9e74">

또한, 현재 `상품에는 항상 하나 이상의 옵션이 있어야 한다.`라는 요구사항을 만족시킬 로직이 존재하지 않는 것 같습니다! 상품을 만들 때 최소 하나의 옵션이 있어야 하고, 상품에서 옵션을 삭제할 때에도 해당 상품이 옵션을 하나밖에 가지고 있지 않다면 예외를 던져주는 기능이 필요해 보이네요!

해당 부분 해결하고 다시 리뷰 요청 부탁드립니다~!



그리고 남겨 주신 질문에서 제가 완벽히 이해하지 못한 것 같은 부분이 있는데요, 
>  만약 다른 테스트에서 스프링 빈으로 등록된 WishService를 사용하게 되면, 이 스프링 빈 객체와는 다른 WishService 객체를 생성해서 쓰게 되는 것이기 때문에 예상치 못한 사이드 이펙트가 생길 수도 있을 것 같다
만약 새로운 서비스 객체를 만들고 이를 테스트에서만 사용한다면, 어떤 사이드 이펙트를 예상하는지 잘 와닿지가 않네요..! (제가 민수님이 말한 의도를 파악하지 못했을 수 있습니다..)

이와는 별개로, 민수님께서 시도해 본 서비스를 Autowiring하는 방법 대신 `@InjectMocks`를 쓰는 방식이 해결책이 될 수도 있겠다는 생각이 듭니다!



```
@DisplayNameGeneration(DisplayNameGenerator.ReplaceUnderscores.class)
@ExtendWith(MockitoExtension.class)
class WishServiceTest {

    @Mock
    ProductRepository productRepository;

    @Mock
    CategoryRepository categoryRepository;

    @InjectMocks
    WishService wishService;
```

minsu-cnu committed:
feat: 상품에는 항상 하나 이상의 옵션이 있어야 한다는 기능 요구 사항 만족

- 상품 삭제 시 연관된 옵션도 연쇄 삭제되도록 설정
- 상품 추가 시, 동시에 기본 옵션을 하나 추가하도록 구현
- 옵션을 삭제 시, 연관된 상품에 대하여 해당 상품에 연관된 옵션이 하나 뿐인 경우 예외 발생 및 처리 (커스텀 예외 구현)

minsu-cnu commented:
질문에 대한 답변 감사드립니다! 사이드 이펙트에 관한 내용은, 사실 어떤 사이드 이펙트가 생길 것이라고 딱 예상한 것은 아니고, 이전에 스프링 빈에 대해 공부할 때, 어떤 한 서비스 객체를 싱글톤으로 딱 하나만 빈으로 등록시켜두고 이를 여러 테스트에서 사용할 때 주입해주는 것이 빈의 이점이라는 것을 알게 되었습니다.

그래서 어떤 서비스 객체가 스프링 빈에도 등록돼있고, 또 어떤 테스트 클래스에서 new 연산자를 통해 별개로 하나 더 만들어서 쓰면 싱글톤 패턴이 퇴색되고, 서로 다른 두 객체를 별개의 테스트에서 각각 쓰고 있는 상황이 되기 때문에, 결과적으로 테스트에 어떠한 예상치 못한 영향이 생기게 될 위험이 있지는 않을까하는 불확실한 고민이 들었어서 질문드리게 되었습니다.

그런데 다시 생각해보니 질문이 좀 추상적이고 불확실한 것에 대한 질문이었던 것 같습니다. 추후에 단위 테스트로 리팩토링을 하게 된다면 그 때 멘토님께서 알려주신 InjectWorks와 new로 객체 생성해주는 방식 두 가지를 모두 적용해보고 직접 경험해보면 좋을 것 같습니다. 조언 감사합니다!!

<br/>
<hr/>
<br/>

말씀해주신 요구 사항을 깜빡하고 놓친 것 같아요.. 죄송합니다..! 바로 해당 요구 사항 구현하여 반영하였습니다.

<br/>
<hr/>
<br/>

제 로컬에서 테스트를 돌렸을 때는 수십번 돌려도 잘 돌아가서.. 어떤 부분이 문제인지 확인이 어려운 상황입니다..! 리뷰 반영 후에도 테스트는 한꺼번에 돌려도 계속 잘 돌아가는데, 혹시나 멘토님 컴퓨터 환경에서 리뷰 반영된 코드로도 여전히 계속 테스트가 실패한다면 알려주시면 감사드리겠습니다!!

jjongwa reviewed (approved):
브랜치 업데이트를 하고 다시 돌리니 이제 다 잘 돌아가네요.. 뭔가 민수님이 커밋을 추가로 하는 도중에 제가 리뷰를 한 것 같네요..!

2단계는 여기서 마무리 하겠습니다! 3단계로 가시죠~~
